### PR TITLE
build: use version.txt file instead of git tags

### LINKED ISSUE #75287: release: fetch tags before linking the tag into the binary, or else make tagging more explicit
The version that CRDB reports itself in telemetry as is determined by the git tags that exist in the local repo at compilation time. See [build/info.go](https://github.com/cockroachdb/cockroach/blob/839b8cf544c7744fe91b1f2000644aef307a87d7/pkg/build/info.go#L29-L31) and the [Makefile](https://github.com/cockroachdb/cockroach/blob/839b8cf544c7744fe91b1f2000644aef307a87d7/Makefile#L1228). I don't know where it happens in the bazel build.

But if somebody is building locally, they might not have all the latest tags. For example I’m seeing data in snowflake from a cluster running on commit 0592999e3a which is a commit on master from August 2021, but it reports itself as v20.1 since it was somebody running locally and that was the most recent tag they had locally.

Can the Makefile and the bazel workflow run `git fetch origin --tags` before linking the tag?

This should be resilient to building offline. If there is an error pulling the tags, it can print a warning message and continue.

---

Update: See @rail's comment below for a different idea to improve the situation.

Jira issue: CRDB-12542

### LINKED ISSUE #91306: build: use a checked-in file to determine binary version instead of git tags
**Describe the problem**

Currently, the binary version is set during build time and determined using `git describe --tags` [1][2][3]. Changing the binary version is done by tagging master and consequently 1) without any corresponding PR and 2) without running our test suite. This means tagging master can have the undesired consequence of breaking tests on master (e.g., version upgrade tests, that now start running new migrations), which is very disruptive.

Ideally, we want these updates to happen by changing a file that is checked in, so that any test failures are caught before the change is merged in the PR that updates the version file.

**Roachtest**

`roachtest` also independently looks at git tags [4]. It should be reading from the same checked-in version file.

**Related issues**

* #75287: missing tags can also lead to wrong binary versions being reported in telemetry. Not relying on git tags would fix that.

[1] https://github.com/cockroachdb/cockroach/blob/master/pkg/build/BUILD.bazel#L21
[2] https://github.com/cockroachdb/cockroach/blob/master/build/bazelutil/stamp.sh#L47
[3] https://github.com/cockroachdb/cockroach/blob/master/pkg/build/info.go#L42
[4] https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/test_registry.go#L202

Jira issue: CRDB-21197

### PR DESCRIPTION
**Pull request by renatolabs**
**build: use version.txt file instead of git tags**
This commit adds a `version.txt` file to the build package. This file
is intended to contain the binary version of cockroach when it is
built at that commit. In order for the binary version to represent
development versions, the revision is appended to the version when a
build is _not_ a release build.

The main benefits of this change are:

* no longer depending on git tag information being available in order
to build cockroach. As a consequence, we should be able to do shallow
clones of cockroach and still be able to build the binary with the
correct version.
* changing the binary version requires a pull request. This means that
tests will need to pass before the version change is merged, stopping
tests that inadvertently rely on a specific version from breaking
master (as they used to when versions relied on git tags).

In terms of the release process itself, the idea is that the version
is bumped whenever a cockroach version is released. So, for example,
when `v23.1.0-alpha.5` is released, the `version.txt` file on master
will be bumped to `v23.1.0-alpha.6`.

This commit also updates `roachtest` so that it uses the `build`
package (and therefore `version.txt`) to determine the binary version
instead of relying on git tags.

**release: update version.txt in `update-versions` command**
This commit makes two changes to the `release` command: introduces the
`set-cockroach-version` subcommand; and uses that command in
`update-versions`.

When a new release is published, we update the roachtest predecessors
map accordingly and, if the version released is the latest one
available, also update Helm charts and Homebrew taps. The updates in
this commit rely on changes to the associated Team City build: we now
also require a "next version" parameter, which will be the value set
in the `version.txt` file on the target branch. The next version needs
to be passed explicitly (at least for now) to avoid the error-prone
process of finding the next version automatically and giving more
flexibility to the release schedule.

Resolves #91306.
Resolves #75287.

Epic: none

Release note: None

### CODE REVIEW

**Comment by renatolabs** [comment]
This is a different approach to remove git tags from the binary after discussion on #95550.

It adds a `version.txt` file that contains the binary version of the next released binary; the PR also includes updates to our release automation so that the version file is updated accordingly when a new release is published.

Release documentation has not been updated yet, but I'll do that once we move forward with the changes here.

**Review by rail — APPROVED** [review]
Love it!
This change will improve reproducibility a lot.
There one thing we should be careful about. After this change we should stop using `master` for the release automation tasks, because it won't work for the older releases. This also means that we should start backporting release automation changes to the release branches.
:lgtm:

> Reviewed 23 of 23 files at r1, 7 of 7 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/97214#-:-NP2uzRkCUupkhaOyLy4:bcmdigh)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @celiala, @herkolategan, @renatolabs, @rickystewart, and @smg260)

___
*[`build/teamcity/internal/cockroach/release/process/update_versions_impl.sh` line 26 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/97214#-NP2r4GV7_dKbXvBoNRr:-NP2r4GV7_dKbXvBoNRs:bnlpn05) ([raw file](https://github.com/cockroachdb/cockroach/blob/c879a68a95d9e8e804a76642ef8d156316229c2d/build/teamcity/internal/cockroach/release/process/update_versions_impl.sh#L26)):*
> ```Shell
>   update-versions \
>   --released-version=$RELEASED_VERSION \
>   --next-version=$NEXT_VERSION \
> ```

Love the explicitness. 
___
*[`pkg/build/info.go` line 59 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/97214#-NP2rSXY-WR8DMC5mUf_:-NP2rSXY-WR8DMC5mUfa:brfv5lf) ([raw file](https://github.com/cockroachdb/cockroach/blob/c879a68a95d9e8e804a76642ef8d156316229c2d/pkg/build/info.go#L59)):*
> ```Go
> }
> 
> func computeBinaryVersion() string {
> ```

Would it be better to leave the same signature, so you don't rely on the global variable? Testing this function would be much easier.
In other words can the signature be `computeBinaryVersion(version string, rev string)`?
___
*[`pkg/cmd/release/update_versions.go` line 155 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/97214#-NP2uhZ727USb-GTgftf:-NP2uhZ727USb-GTgftg:b-83sjoh) ([raw file](https://github.com/cockroachdb/cockroach/blob/c879a68a95d9e8e804a76642ef8d156316229c2d/pkg/cmd/release/update_versions.go#L155)):*
> ```Go
> 	for _, repo := range repos {
> 		dest := path.Join(dir, repo.name())
> 		// We avoid creating duplicated PRs to allow this command to be
> ```

Nice!

File: build/bazelutil/stamp.sh [hunk 1]
```diff
@@ -41,15 +41,6 @@ else
     shift 1
 fi
 
-# Handle build-tag.
-if [ -z "${1+x}" ]
-then
-    BUILD_TAG=$(git describe --tags --dirty --match=v[0-9]* 2> /dev/null || git rev-parse --short HEAD;)
```
Comment (rickystewart) [inline build/bazelutil/stamp.sh:47]: You must delete the line `BUILD_TAG ${BUILD_TAG-}` from the bottom of the script since we're un-defining this variable.

Comment (rickystewart) [reply build/bazelutil/stamp.sh:47]: Also remove the `build-tag` argument from the documentation above (line 12)

**Review by rickystewart — CHANGES_REQUESTED** [review]
The release code needs to be updated accordingly. Can you apply the following patch?

```
diff --git a/build/bazelutil/stamp.sh b/build/bazelutil/stamp.sh
index a76b504fcb0..0966353da8e 100755
--- a/build/bazelutil/stamp.sh
+++ b/build/bazelutil/stamp.sh
@@ -9,7 +9,6 @@
 # for a development build.
 #  target-triple: defaults to the value of `cc -dumpmachine`
 #  build-channel: defaults to `unknown`, but can be `official-binary`
-#  build-tag: defaults to a value that is gleaned from `git rev-parse`
 #  build-type: defaults to `development`, but can be `release`
 
 set -euo pipefail
diff --git a/pkg/cmd/publish-provisional-artifacts/main_test.go b/pkg/cmd/publish-provisional-artifacts/main_test.go
index 60c6df82b7f..3066187c8b3 100644
--- a/pkg/cmd/publish-provisional-artifacts/main_test.go
+++ b/pkg/cmd/publish-provisional-artifacts/main_test.go
@@ -174,26 +174,26 @@ func TestProvisional(t *testing.T) {
 			},
 			expectedCmds: []string{
 				"env=[] args=bazel build //pkg/cmd/cockroach //pkg/cmd/cockroach-sql //c-deps:libgeos " +
-					"'--workspace_status_command=./build/bazelutil/stamp.sh x86_64-pc-linux-gnu official-binary v0.0.1-alpha release' -c opt --config=ci --config=force_build_cdeps --config=with_ui --config=crosslinuxbase",
+					"'--workspace_status_command=./build/bazelutil/stamp.sh x86_64-pc-linux-gnu official-binary release' -c opt --config=ci --config=force_build_cdeps --config=with_ui --config=crosslinuxbase",
 				"env=[] args=bazel info bazel-bin -c opt --config=ci --config=force_build_cdeps --config=with_ui --config=crosslinuxbase",
 				"env=[MALLOC_CONF=prof:true] args=./cockroach.linux-2.6.32-gnu-amd64 version",
 				"env=[] args=ldd ./cockroach.linux-2.6.32-gnu-amd64",
 				"env=[] args=bazel run @go_sdk//:bin/go -- tool nm ./cockroach.linux-2.6.32-gnu-amd64",
-				"env=[] args=bazel build //pkg/cmd/cockroach //pkg/cmd/cockroach-sql //c-deps:libgeos '--workspace_status_command=./build/bazelutil/stamp.sh x86_64-pc-linux-gnu official-binary v0.0.1-alpha release' -c opt --config=ci --config=force_build_cdeps --config=with_ui --config=crosslinuxfipsbase",
+				"env=[] args=bazel build //pkg/cmd/cockroach //pkg/cmd/cockroach-sql //c-deps:libgeos '--workspace_status_command=./build/bazelutil/stamp.sh x86_64-pc-linux-gnu official-binary release' -c opt --config=ci --config=force_build_cdeps --config=with_ui --config=crosslinuxfipsbase",
 				"env=[] args=bazel info bazel-bin -c opt --config=ci --config=force_build_cdeps --config=with_ui --config=crosslinuxfipsbase",
 				"env=[MALLOC_CONF=prof:true] args=./cockroach.linux-2.6.32-gnu-amd64-fips version",
 				"env=[] args=ldd ./cockroach.linux-2.6.32-gnu-amd64-fips",
 				"env=[] args=bazel run @go_sdk//:bin/go -- tool nm ./cockroach.linux-2.6.32-gnu-amd64-fips",
 				"env=[] args=bazel build //pkg/cmd/cockroach //pkg/cmd/cockroach-sql //c-deps:libgeos " +
-					"'--workspace_status_command=./build/bazelutil/stamp.sh x86_64-apple-darwin19 official-binary v0.0.1-alpha release' -c opt --config=ci --config=force_build_cdeps --config=with_ui --config=crossmacosbase",
+					"'--workspace_status_command=./build/bazelutil/stamp.sh x86_64-apple-darwin19 official-binary release' -c opt --config=ci --config=force_build_cdeps --config=with_ui --config=crossmacosbase",
 				"env=[] args=bazel info bazel-bin -c opt --config=ci --config=force_build_cdeps --config=with_ui --config=crossmacosbase",
-				"env=[] args=bazel build //pkg/cmd/cockroach //pkg/cmd/cockroach-sql '--workspace_status_command=./build/bazelutil/stamp.sh aarch64-apple-darwin21.2 official-binary v0.0.1-alpha release' -c opt --config=ci --config=force_build_cdeps --config=with_ui --config=crossmacosarmbase",
+				"env=[] args=bazel build //pkg/cmd/cockroach //pkg/cmd/cockroach-sql '--workspace_status_command=./build/bazelutil/stamp.sh aarch64-apple-darwin21.2 official-binary release' -c opt --config=ci --config=force_build_cdeps --config=with_ui --config=crossmacosarmbase",
 				"env=[] args=bazel info bazel-bin -c opt --config=ci --config=force_build_cdeps --config=with_ui --config=crossmacosarmbase",
 				"env=[] args=bazel build //pkg/cmd/cockroach //pkg/cmd/cockroach-sql //c-deps:libgeos " +
 					"'--workspace_status_command=." +
-					"/build/bazelutil/stamp.sh x86_64-w64-mingw32 official-binary v0.0.1-alpha release' -c opt --config=ci --config=force_build_cdeps --config=with_ui --config=crosswindowsbase",
+					"/build/bazelutil/stamp.sh x86_64-w64-mingw32 official-binary release' -c opt --config=ci --config=force_build_cdeps --config=with_ui --config=crosswindowsbase",
 				"env=[] args=bazel info bazel-bin -c opt --config=ci --config=force_build_cdeps --config=with_ui --config=crosswindowsbase",
-				"env=[] args=bazel build //pkg/cmd/cockroach //pkg/cmd/cockroach-sql //c-deps:libgeos '--workspace_status_command=./build/bazelutil/stamp.sh aarch64-unknown-linux-gnu official-binary v0.0.1-alpha release' -c opt --config=ci --config=force_build_cdeps --config=with_ui --config=crosslinuxarmbase",
+				"env=[] args=bazel build //pkg/cmd/cockroach //pkg/cmd/cockroach-sql //c-deps:libgeos '--workspace_status_command=./build/bazelutil/stamp.sh aarch64-unknown-linux-gnu official-binary release' -c opt --config=ci --config=force_build_cdeps --config=with_ui --config=crosslinuxarmbase",
 				"env=[] args=bazel info bazel-bin -c opt --config=ci --config=force_build_cdeps --config=with_ui --config=crosslinuxarmbase",
 			},
 			expectedGets: nil,
diff --git a/pkg/release/build.go b/pkg/release/build.go
index 8eacf3283a3..769755d0a6f 100644
--- a/pkg/release/build.go
+++ b/pkg/release/build.go
@@ -152,7 +152,7 @@ func MakeRelease(platform Platform, opts BuildOptions, pkgDir string) error {
 		if opts.BuildTag == "" {
 			return errors.Newf("must set BuildTag if Release is set")
 		}
-		buildArgs = append(buildArgs, fmt.Sprintf("--workspace_status_command=./build/bazelutil/stamp.sh %s official-binary %s release", targetTriple, opts.BuildTag))
+		buildArgs = append(buildArgs, fmt.Sprintf("--workspace_status_command=./build/bazelutil/stamp.sh %s official-binary release", targetTriple))
 	} else {
 		if opts.BuildTag != "" {
 			return errors.Newf("cannot set BuildTag if Release is not set")

```

**Review by renatolabs — COMMENTED** [review]
Done, thanks for the patch!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/97214#-:-NPJL2oa33Qw8j4Iq5Ue:b-863vs5)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @celiala, @herkolategan, @rail, and @smg260)

*[a discussion](https://reviewable.io/reviews/cockroachdb/cockroach/97214#-NPJL7FPEGuDVVvP84dG:-NPJL7FPEGuDVVvP84dH:b-lz7t30) (no related file):*
> After this change we should stop using master for the release automation tasks, because it won't work for the older releases.

@rail could you elaborate what you meant here, please? Is it because the path to roachtest's `predecessors.json` is different on master vs release-22.2 so `update-versions` won't work? This is something we should fix if we're using `update-versions` on 22.2.* releases, but it was already an issue before these changes, so I suspect you meant something else.

My understanding is that `version.txt` shouldn't interfere with 22.2 releases as that's file is untracked there.

___
*[a discussion](https://reviewable.io/reviews/cockroachdb/cockroach/97214#-NPJMKO-3WHxqBbBzMkV:-NPJMKO-3WHxqBbBzMkW:b116ihy) (no related file):*
This is RFAL, I think I addressed all comments.

___
___
*[`pkg/build/info.go` line 59 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/97214#-NP2rSXY-WR8DMC5mUf_:-NPJKiiiBJkwB2TCLPfr:b-m0auy8) ([raw file](https://github.com/cockroachdb/cockroach/blob/c879a68a95d9e8e804a76642ef8d156316229c2d/pkg/build/info.go#L59)):*
<details><summary><i>Previously, rail (Rail Aliiev) wrote…</i></summary><blockquote>

Would it be better to leave the same signature, so you don't rely on the global variable? Testing this function would be much easier.
In other words can the signature be `computeBinaryVersion(version string, rev string)`?
</blockquote></details>

Makes sense, I followed your suggestion and added parameters to the function. I ended up adding tests for it too to make sure we don't break the current logic by accident.

**Comment by rail** [comment]
> @rail could you elaborate what you meant here, please? Is it because the path to roachtest's `predecessors.json` is different on master vs release-22.2 so `update-versions` won't work? This is something we should fix if we're using `update-versions` on 22.2.* releases, but it was already an issue before these changes, so I suspect you meant something else.

Also we would need to bump the version.txt file, which is not on the branches. We can try keeping the master version of pkg/cmd/release work for all branches, but it may become a rabbit hole in the future. 

 I'm totally fine running our release automation using branches, this is something we expected. At some point the release process will be diverged.

**Comment by renatolabs** [comment]
> we would need to bump the version.txt file, which is not on the branches

I think that should be fine as the file would be ignored when we run `git commit` (as it's an untracked file in release branches). That said, I agree it's more sane to run the release branch version of `pkg/cmd/release`.

**Review by rickystewart — APPROVED** [review]

**Comment by rickystewart** [comment]
Thanks for doing this Renato, this is a great change. This is *far* superior to what we were doing before, and much easier to understand. It'll make it easier to make some necessary improvements on the Bazel side as well.

**Review by rail — APPROVED** [review]
:lgtm:

> Reviewed 14 of 14 files at r3, 5 of 7 files at r4, 4 of 4 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/97214#-:-NPYYJbHAYhmWKmEMBjI:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @celiala, @herkolategan, and @smg260)

**Comment by renatolabs** [comment]
TFTRs!

bors r=rail,rickystewart

💥
